# Coalescent simulation {#coalescentsim}

This chapter deals with the technical details of coalescent simulation.  We have seen some simulations run in Chapter \@ref(treeseqs), but we have not yet defined what we mean by a coalescent simulation.  As it happens, there is no one coalescent simulation.  Rather, there are many families of models for generating the genealogies of samples.  We will start with a brief description of Kingman's continuous-time labelled coalescent, which is the "standard" model of the field, discovered independently by @Kingman1982-cq, @Tajima1983-it, and @Hudson1983-kn.  The key features that we will describe are the structures of trees and the measurement of time. Many excellent reviews already exist on this flavor of the coalescent (@Wakeley2008-hd, @Rosenberg2002-ac, @Hudson1990-ff).

## Labelled trees

If we look at the trees shown in previous chapters, there are several important features to keep in mind. For example, in Figure \@ref(fig:drawtree), there is a strict ordering of node times (and their corresponding labels/indexes) from the present to the past.  Further, every node is labelled, and ancestral nodes are labelled according to their set of children, who are in turn labelled by their set of children, etc., until we reach the present day sample, which we may think of as a separate labelling (the set of modern-day alleles that are the descendants of a specific ancestral node).

In trees like these, a node is a specific genome at a specific point in time.  Thinking in terms of edges (Section \@ref(treeseqdetails)), a node node represents a set of parent/child relationships over a specific genomic interval.


## The continuous time approximation


## Some history

TBD: ARG/HRG/PSMC/TS

## Hudson's linear-time algorithm

In a 1990 review article [@Hudson1990-ff], Dick Hudson published the C
code for a linear time algorithm to simulate gene genealogies under the Kingman 
coalescent. His method works by realizing that for a sample of $n$ chromosomes,
the final tree with have $2n - 1$ nodes that we may label $[0, 2n)$. 
The first $n$ nodes correspond to the present day sample, and the remaining
$n-1$ are the possible ancestors.  For example, if $n = 4$, we have:

```{python}
import numpy as np
n = 4
node_labels = np.arange(2*n - 1, dtype=np.int32)
print(node_labels)
```

Our samples are:

```{python}
print(node_labels[:n])
```

Our ancestral nodes are:

```{python}
print(node_labels[n:])
```

More generally, at any time in the past where $i$ lineages remain,
the first $i$ nodes are represent the sample, and the remaining nodes
are their possible parents. We wish to take a random pair from the first $i$
nodes and assign as their parent the first of the remaining possible ancestral nodes.

The algorithm to track the IDs of the remaining sample nodes proceeds as follows:

a. *Choose first sample node* Choose a value `p` uniformly from $[1, i)$ and record the node label at that position, *e.g.* set `c1 = node_labels[p]`.
b. *Update the sample list* Set `node_labels[c1] = node_labels[i-1]`, removing `c1` from the list of samples remaining.
c. *Choose second sample node* Choose `p` uniformly from $[i, i-1)$, and set `c2 = node_labels[p]`. The swap in step B prevents `c1` from being chosen again.
d. *Update the sample list* Set `node_labels[c2] = node_labels[2*nsam - i]`. 
e. Set `i = i - 1`.  If `i == 1`, stop, otherwise return to A.

Step D is the tricky one.  It moves the first of the remaining ancestor IDs into the last position of
of the remaining sample IDs. At the end of this step, `node_labels[c2]`'s value is the ID of the
parent of `c1` and `c2`.

The above algorithm is very straightforward to write out in vanilla
Python.  We also generate the tree corresponding to the simulated history.
The tree is represented as parent indexes. We also generate the time of 
the coalescent events.

```{python hudson1990py, code=readLines("chapters/coalescent/listings/hudson1990.py")}
```

Let's run the function, and look at the resulting tree structure:

```{python}
np.random.seed(23456)
tree, times = h1990(4)
for i in range(7):
   print("node {}, parent {}, time {}".format(i,tree[i],times[i]))
```

The above code works and is very simple.  
The method is limited, though, as we have represented the tree using the classic
"list of indexes of parents" approach (CITATION).  A better approach would
be to modify the above algorithm to generate tree sequence data structures 
(Chapter \@ref(treeseqs)), which we do in the following code block,
which is part of the source code distributed with this book:

```{python, code=readLines("../simbook/coalescent/hudson1990.py"), eval=F}
```

The previous code block was used to generate Figure \@ref(fig:h1990fig).

```{r h1990fig, echo=FALSE, fig.cap="The distribution of the time to the most recent common ancestor (TMRCA) and the total time on the tree (TTOT) for a sample size of $n = 50$. The distributions are based on 10,000 simulation replicates."}
knitr::include_graphics("hudson1990.png")
```

## The coalescent with migration   

In this section, we extend the algorithm to allow for symmetric migration between two discrete demes.  This is the
standard "island model" of migration (CITATION). 

With population structure, two lineages may only coalesce if they are both present in the same deme.  Migration events
change the label of a deme.  Thus, we must track the current number of lineages in each deme.  This additional
requirement substantially increases the complexity of the simulation, as we now have the following possible events:

1. Coalescence in deme 0
2. Coalescence in deme 1
3. Migration from deme 0 to deme 1
4. Migration from deme 1 to deme 0

For simplicity, we assume:

1. The migration rate from deme zero to one equals the rate from one to zero.
2. The effective size each deme is the same.

Define $m$ as the rate (per $2N$ generations) at which a lineage migrates to the other deme, where $N$ is the effective size of a deme. 
At any time in the past, there are $n_0$ lineages remaining in deme zero and $n_1$ remaining in deme one.
The migration rate per $2N$ generation from deme zero to deme one is $mn_0$, and the rate in the reverse
direction is $mn_1$.  The total migration rate is therefore $M=m(n_0+n_1)$, and the mean time to the next migration event is
exponentially-distributed with mean $1/M$.  Likewise, the rates of coalescence in the demes are
${n_0 \choose 2}$ and ${n_1 \choose 2}$, respectively, and the total rate of coalescence is the sum of the
two rates.

To get the time to the next event, we have several options. For example, we could draw the next time for each of the four possible events
and the minimum value determines the event.  This method would give the correct results but would be relatively
inefficient, as exponential deviates are one of the more expensive to generate.

A more efficient method is to generate the time to the next migration event (mean $1/M$ time units), and the time
to the next coalescent event (mean $1/\left[{n_0 \choose 2} + {n_1 \choose 2}\right]$ time units).  If the time to the next
migration event is the smaller of the two, we chose the deme proportional to the sample size within each deme.  If the
time to the next coalescent event is the smaller value, we choose deme proportional to their rates of coalescence.

The following listing shows the algorithm for the case of two demes:

```{python, code=readLines("../simbook/coalescent/migration.py"), eval=F}
```

Let's look at some output:

```{python migtree, code=readLines("chapters/coalescent/listings/migration_tree.py")}
```

In the above tree, a mutation on any internal node leading to descendants in both demes will be a shared polymorphism.
A mutation on a node leading only to samples from one population results in a private polymorphism or a fixed difference 
between demes.  The relative number of shared, private, and fixed mutations determines the value of summaries of
differentiation such as $F_{st}$ and the $f$ statistics from [@Patterson2012-sw].

Figure \@ref(fig:migration) compares the results of our implementation to `msprime`:

```{r migration, echo=FALSE, fig.cap="The empirical cumulative density function (ECDF) of TMRCA and the $f_2$ statistic [@Patterson2012-sw] in a model of symmetric migration between two demes of equal size.  The initial sample sizes in each deme are 10 and 5.  The total migration rate (per lineage per 2N generations) is 0.5, and migration rates are symmetric between demes.  The comparison to `msprime` is based on 10,000 replicates and accounts for the difference in time scale between the two implementations. The sudden drop to zero at the right of each panel is a side-effect of plotting ECDF functions using matplotlib."}
knitr::include_graphics("migration.png")
```

**Discuss how multiple demes, changing deme size, etc., all affect things.**
